File: AdventOfCode2019/03/day3.hs
import Data.List (elemIndex)
import Data.Maybe (isNothing, catMaybes)
import Input (wire1, wire2)
import qualified Data.Map as M

move :: (Integral a) => (a, a) -> (Char, a) -> [(a, a)]
move (x, y) (dir, steps)
  | dir == 'L' = [(x - i, y) | i <- [1 .. steps]]
  | dir == 'R' = [(x + i, y) | i <- [1 .. steps]]
  | dir == 'U' = [(x, y + i) | i <- [1 .. steps]]
  | dir == 'D' = [(x, y - i) | i <- [1 .. steps]]
  | otherwise = error "Wrong rule"

pathPoints :: Integral a => (a, a) -> [(Char, a)] -> [(a, a)]
pathPoints _ [] = []
pathPoints pos (x : xs) = let path = move pos x in path ++ pathPoints (last path) xs

fullPath x = M.fromListWith min . flip zip [1 ..] $ pathPoints (0, 0) x

wire1Path = fullPath wire1

wire2Path = fullPath wire2

intersect :: Eq a => [a] -> [a] -> [a]
intersect [] _ = []
intersect _ [] = []
intersect xs ys = filter (`elem` xs) ys

intersection = foldr1 M.intersection [wire1Path, wire2Path]

manhattanDistance (x, y) = abs x + abs y

part1 = minimum . map manhattanDistance . M.keys $ M.intersection wire1Path wire2Path

part2 = minimum . M.elems $ M.intersectionWith (+) wire1Path wire2Path

main = do
  print intersection
  print part1
  print part2

File: AdventOfCode2019/03/day3.js
var input = require("./input.js");

const directionToVectorMap = {
  L: [-1, 0],
  R: [1, 0],
  U: [0, 1],
  D: [0, -1],
};

const grid = {
  values: {}, // x,y => {pathName: delay}
  intersections: [],
  draw: function (name, path) {
    let [x, y] = [0, 0];
    let delay = 1;
    path.forEach(([direction, amount]) => {
      const [dx, dy] = directionToVectorMap[direction];
      for (let i = 1; i <= amount; i++, delay++) {
        [x, y] = [x + dx, y + dy];
        const key = `${x},${y}`;
        const currentCell = (this.values[key] = this.values[key] || {});
        if (currentCell[name]) {
          continue;
        }

        if (Object.keys(currentCell).length > 0) {
          this.intersections.push([x, y]);
        }

        currentCell[name] = delay;
      }
    });
  },
  closestIntersection: function (distanceFunc) {
    const distances = grid.intersections.map(distanceFunc);
    const min = Math.min.apply(null, distances);
    return min;
  },
};

grid.draw(0, input.wire1);
grid.draw(1, input.wire2);

const manhattanDistance = ([x, y]) => Math.abs(x) + Math.abs(y);
const part1 = grid.closestIntersection(manhattanDistance);
console.log(part1);

const getTiming = (pos) => {
  const cell = grid.values[pos.toString()];
  return cell[0] + cell[1];
};
const part2 = grid.closestIntersection(getTiming);
console.log(part2);
